lib/PostsIterator.php

```diff
@@ -6,11 +6,44 @@
  * Class PostsIterator
  */
 class PostsIterator extends \ArrayIterator {
-
+	/**
+	 * Prepares the state before working on a post.
+	 *
+	 * Calls the setup function of the current post to setup post data.
+	 * Before starting the loop, it will call the 'loop_start' hook to improve compatibility.
+	 *
+	 * @return mixed
+	 */
 	public function current() {
-		global $post;
+		// Fire action when the loop has just started.
+		if ( 0 === $this->key() ) {
+			do_action_ref_array( 'loop_start', array( &$GLOBALS['wp_query'] ) );
+		}
+
 		$post = parent::current();
+		$post->setup();
+
 		return $post;
 	}
 
+	/**
+	 * Cleans up state before advancing to next item.
+	 *
+	 * In the last run of a loop through posts, it will call the 'loop_end' hook to improve
+	 * compatibility.
+	 *
+	 * @since 2.0.0
+	 */
+	public function next() {
+		$post = $this->current();
+		$post->teardown();
+
+		// Fire action when the loop has ended.
+		if ( $this->key() === $this->count() - 1 ) {
+			do_action_ref_array( 'loop_end', array( &$GLOBALS['wp_query'] ) );
```

Comment:
@gchtr I think we still miss to reset the `$wp_query->in_the_loop` here? Just had this problem in a project of mine. We could also just call `have_posts()` to reset it. Any thoughts?

Reply:
@pascalknecht Yes, I forgot about that! But shouldn’t it be in `Timber\Post::teardown()`, when we already set `$wp_query->in_the_loop` to `true` in `Timber\Post::setup()`?

We could call `have_posts()`, but then we should also handle `$wp_query->current_post` so that `have_posts()` acts the way it should. We could also let it do the call the `loop_end` action.

Reply:
This is the only outstanding failing test. Is it too hacky to simply do...
```php
global $wp_query;
$wp_query->in_the_loop = false;
```
... at this point in `PostsIterator.php`? currently the `loop_end` action doesn't seem to change the value for `$wp_query->in_the_loop`

Reply:
Yes, I added the test, but not the code yet, because I wasn’t sure about the cleanest solution.

@jarednova I don’t think it’s too hacky. But I was thinking about where to put it. We set `$wp_query->in_the_loop` to `true` in `Timber\Post::setup()`. Shouldn’t we add the counterpart in `Timber\Post::teardown()` then, instead of the `PostsIterator` class? Or is this more of a philosophical question?

Reply:
I think it exists in the realm of the philosophical. Two will enter. One will win, one will lose. Undoubtedly there should be a comment in the "loser" (`Post` vs. `PostsIterator`) that points to where it ultimately lives in the "winner." Once you establish that, I think it's your prerogative on which makes the most sense (I have very convincing arguments going in my head for either)